src/facade/arg_parser.h

```diff
@@ -0,0 +1,176 @@
+// Copyright 2023, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#pragma once
+
+#include <absl/strings/match.h>
+
+#include <string_view>
+
+#include "facade/error.h"
+#include "facade/facade_types.h"
+
+namespace facade {
+
+struct ArgumentParser {
+  struct NextProxy;
+  friend struct NextProxy;
+  struct CheckProxy;
+  friend struct CheckProxy;
+
+  enum ErrorType {
+    OUT_OF_BOUNDS,
+    SHORT_OPT_TAIL,
+    INVALID_INT,
+    INVALID_CASES,
+  };
+
+  struct NextProxy {
+    operator std::string_view() {
+      return ToSV(parser_->args_[idx_]);
+    }
+
+    operator std::string() {
+      return std::string{operator std::string_view()};
+    }
+
+    template <typename T> operator T() {
+      T out;
+      if (absl::SimpleAtoi(operator std::string_view(), &out))
+        return out;
+      parser_->Report(INVALID_INT, idx_);
+      return T{0};
+    }
+
+    template <typename T> T Cases(std::initializer_list<std::pair<std::string_view, T>> values) {
+      std::string_view arg = operator std::string_view();
+      for (const auto& [tag, value] : values) {
+        if (absl::EqualsIgnoreCase(tag, arg))
+          return value;
+      }
+      parser_->Report(INVALID_CASES, idx_);
+      return T{};
+    }
+
+   private:
+    friend struct ArgumentParser;
+
+    NextProxy(ArgumentParser* parser, size_t idx) : parser_{parser}, idx_{idx} {
+    }
+
+    ArgumentParser* parser_;
+    size_t idx_;
+  };
+
+  struct CheckProxy {
+    operator bool() {
+      if (idx_ >= parser_->args_.size())
+        return false;
+
+      std::string_view arg = ToSV(parser_->args_[idx_]);
+      if (arg != tag_)
+        return false;
+
+      if (idx_ + expect_tail_ >= parser_->args_.size()) {
+        parser_->Report(SHORT_OPT_TAIL, idx_);
+        return false;
+      }
+
+      parser_->cur_i_++;
+
+      if (next_upper_)
+        parser_->ToUpper(idx_ + expect_tail_ + 1);
+
+      return true;
+    }
+
+    CheckProxy& ExpectTail(size_t tail) {
+      expect_tail_ = tail;
+      return *this;
+    }
+
+    CheckProxy& NextUpper() {
+      next_upper_ = true;
+      return *this;
+    }
+
+   private:
+    friend struct ArgumentParser;
+
+    CheckProxy(ArgumentParser* parser, std::string_view tag, size_t idx)
+        : parser_{parser}, tag_{tag}, idx_{idx} {
+    }
+
+    ArgumentParser* parser_;
+    std::string_view tag_;
+    size_t idx_;
+    size_t expect_tail_ = 0;
+    bool next_upper_ = false;
+  };
+
+  struct ErrorInfo {
+    ErrorType type;
+    size_t index;
+
+    ErrorReply MakeReply() {
+      return ErrorReply{kSyntaxErr};  // switch here
+    }
+  };
+
+ public:
+  ArgumentParser(CmdArgList args) : args_{args} {
+  }
+
+  NextProxy Peek() {
+    return Next(0);
+  }
+
+  NextProxy Next(size_t step = 1) {
+    if (cur_i_ >= args_.size())
```

Comment:
You report an error and then you return an object that will access out-of-bounds when cast to string_view. Can this return `io::Result<NextProxy>` or something?

Reply:
Was just checking it works, my idea is to return an empty string_view in this case

Reply:
My issue with parsing code is that I need to do a lot of checks with returns.

See this example 
```c++
if (ArgS(args, i) == "LIMIT") {
      if (i + 2 >= args.size()) {
        (*cntx)->SendError(kSyntaxErr);
        return nullopt;
      }
      if (!absl::SimpleAtoi(ArgS(args, i + 1), &limit_offset) ||
          !absl::SimpleAtoi(ArgS(args, i + 2), &limit_total)) {
        (*cntx)->SendError(kInvalidIntErr);
        return nullopt;
      }
      i += 2;
      continue;
}
```

It becomes just

```c++
if (parser.Check("LIMIT").ExpectTail(2)) {
      limit_offset = parser.Next().Int<size_t>();
      limit_total = parser.Next().Int<size_t>();
      continue;
}
```

because I saved error reporting. It's done at the very end. Because parsing code performs usually no actions, its safe to let it run a little. 

What I thought of is using exceptions? Isn't it the tool that was invented exactly for this use-case. Afaik their overhead for non-throwing invocations is almost negligible - and its also a very handy tool of execution control.

So the following is possible
```c++
optional<Schema> ParseSchemaOrReply(args, cntx) try {
  // use parser
  return schema;
} catch (exception) {
  cntx->SendError(execption)
  return nullopt;
}

```

Reply:
I wouldn't want to put exceptions here..
What about turning the bounds check just into an assert and assume the bounds check is done externally? As is done in your example already.

Reply:
Its done in my example, but not in all places

Reply:
Ok. I think without out-of-bounds reads and with the error-checking asserts it's reasonably safe.